### PR TITLE
fix prepend StackOverflow issue

### LINKED ISSUE #54711: Stack overflow during gc_read_stack on win32
Saw in a few places. One example here: https://buildkite.com/julialang/julia-master/builds/37111#018feb4c-a436-49ef-b2a3-63213aa05f67.

### PR DESCRIPTION
**Pull request by vtjnash**
Attempt to fix #54711
Test introduced by #36227

### CODE REVIEW

File: base/array.jl [hunk 1]
```diff
@@ -1316,8 +1316,7 @@ end
 
 append!(a::AbstractVector, iter) = _append!(a, IteratorSize(iter), iter)
 push!(a::AbstractVector, iter...) = append!(a, iter)
-
-append!(a::AbstractVector, iter...) = foldl(append!, iter, init=a)
+append!(a::AbstractVector, iter...) = for v in iter; append!(a, v); end
```
Comment (aviatesk) [inline base/array.jl:1319]:
```suggestion
append!(a::AbstractVector, iter...) = (for v in iter; append!(a, v); end; return a)
```

File: base/array.jl [hunk 2]
```diff
@@ -1376,10 +1375,9 @@ function prepend!(a::Vector{T}, items::Union{AbstractVector{<:T},Tuple}) where T
     return a
 end
 
-prepend!(a::Vector, iter) = _prepend!(a, IteratorSize(iter), iter)
-pushfirst!(a::Vector, iter...) = prepend!(a, iter)
-
-prepend!(a::AbstractVector, iter...) = foldr((v, a) -> prepend!(a, v), iter, init=a)
+prepend!(a::AbstractVector, iter) = _prepend!(a, IteratorSize(iter), iter)
+pushfirst!(a::AbstractVector, iter...) = prepend!(a, iter)
+prepend!(a::AbstractVector, iter...) = for v in iter; prepend!(a, v); end
```
Comment (aviatesk) [inline base/array.jl:1380]:
```suggestion
prepend!(a::AbstractVector, iter...) = (for v in iter; prepend!(a, v); end; return a)
```

File: base/array.jl [hunk 3]
```diff
@@ -1376,10 +1375,9 @@ function prepend!(a::Vector{T}, items::Union{AbstractVector{<:T},Tuple}) where T
     return a
 end
 
-prepend!(a::Vector, iter) = _prepend!(a, IteratorSize(iter), iter)
-pushfirst!(a::Vector, iter...) = prepend!(a, iter)
-
-prepend!(a::AbstractVector, iter...) = foldr((v, a) -> prepend!(a, v), iter, init=a)
+prepend!(a::AbstractVector, iter) = _prepend!(a, IteratorSize(iter), iter)
+pushfirst!(a::AbstractVector, iter...) = prepend!(a, iter)
+prepend!(a::AbstractVector, iter...) = (for v in iter; prepend!(a, v); end; return a)
```
Comment (aviatesk) [inline base/array.jl:1380]:
```suggestion
prepend!(a::AbstractVector, iter...) = (for i = length(iter):-1:1; prepend!(a, iter[i]); end; return a)
```

File: base/array.jl [hunk 4]
```diff
@@ -1376,10 +1375,9 @@ function prepend!(a::Vector{T}, items::Union{AbstractVector{<:T},Tuple}) where T
     return a
 end
 
-prepend!(a::Vector, iter) = _prepend!(a, IteratorSize(iter), iter)
-pushfirst!(a::Vector, iter...) = prepend!(a, iter)
-
-prepend!(a::AbstractVector, iter...) = foldr((v, a) -> prepend!(a, v), iter, init=a)
+prepend!(a::AbstractVector, iter) = _prepend!(a, IteratorSize(iter), iter)
+pushfirst!(a::AbstractVector, iter...) = prepend!(a, iter)
+prepend!(a::AbstractVector, iter...) = (for i = length(iter):-1:1; prepend!(a, iter[i]); end; return a)
```
Comment (vtjnash) [inline base/array.jl:1380]: alternatively? I guess these should generate the same code though?
```suggestion
prepend!(a::AbstractVector, iter...) = (for i = reverse(iter); prepend!(a, i); end; return a)
```

**Comment by staticfloat** [comment]
@vtjnash We are seeing that https://github.com/JuliaLang/julia/issues/52434 no longer seems to happen after this was merged.  Could this have possibly fixed that issue on aarch64 as well?

**Comment by giordano** [comment]
The kernel issue was related only to the `SharedArrays` issue, the discussion in the thread mixed two separate issues.

**Comment by matthias314** [comment]
What was the reason to replace `foldl` and `foldr` by loops? It seems to me that this has made `append!` and in particular `preprend` slower:
```
julia> t = (1:2, 3:4, 5:6, 7:8);

julia> f1(v, t...) = foldl((v, w) -> append!(v, w), t; init = v);

julia> @b [1,2,3] append!(_, $t...) evals = 1
148.000 ns (6 allocs: 592 bytes)

julia> @b [1,2,3] f1(_, $t...) evals = 1
123.000 ns (2 allocs: 464 bytes)

julia> f2(v, t...) = foldr((w, v) -> prepend!(v, w), t; init = v);

julia> @b [1,2,3] prepend!(_, $t...) evals = 1
417.000 ns (12 allocs: 880 bytes)

julia> @b [1,2,3] f2(_, $t...) evals = 1
128.000 ns (2 allocs: 464 bytes)
```

**Pull request merged**